### ISSUE TITLE
Identify patients with Simprints

### ISSUE THREAD

**Issue by abbyad** [medic/medic-webapp#3511]
Simprints is working on building low-cost, rugged fingerprint scanners for frontline workers. We can integrate use of these scanners into health worker workflows to identify patients. This can be helpful for a variety of projects, such as those with large number of patients that attend a clinic, ones with community campaigns, and ones where patient identification using names alone is not reliable.

There are two Simprints operations of interest right now: 
- **Register:** Adding a patient's biometric info to their profile
- **Identify:** Finding a patient's profile using their biometric info

Here are some mockups showing both of these functions:

## Register
![medic mobile--simprints workflow mockup 1](https://cloud.githubusercontent.com/assets/789512/26496566/09ca0c9a-41f6-11e7-90dd-9f4550245442.png)

## Identify
![medic mobile--simprints workflow mockup 2](https://cloud.githubusercontent.com/assets/789512/26496569/0c30b538-41f6-11e7-8b51-3be762be8f7a.png)

![medic mobile--simprints workflow mockup 3](https://cloud.githubusercontent.com/assets/789512/26496571/0f8098de-41f6-11e7-88eb-9e7d96912980.png)

**Comment by garethbowen** [comment]
Code review please @alxndrsn (four PRs, one in each of webapp, api, sentinel, and android).

I haven't tested the android app as I don't have a simprints device - if you've got one running a quick test on this branch would be really useful.

The main change IMO was using APIs for communication between the webapp and the android container rather than injecting straight into the DOM. The rest is just trying to get more reuse, internationalising everything, and cleaning stuff up.

**Comment by alxndrsn** [comment]
@garethbowen without simprint hardware device, you can test with this android app: https://github.com/alxndrsn/mocksimprints

**Comment by alxndrsn** [comment]
If the webapp PR is at https://github.com/medic/medic-webapp/pull/3510, please rebase onto master for review.

**Comment by garethbowen** [comment]
Thanks @alxndrsn - I've made some of the changes from your code review and queried some of the others. Can you please review these again?

https://github.com/medic/medic-webapp/pull/3774
https://github.com/medic/medic-android/pull/37

**Comment by ngaruko** [comment]
Hi @abbyad . Did you want me to AT this using the android app that Alex suggested above or did you wanna AT this with the real hardware?

**Comment by sglangevin** [comment]
Assigning to @abbyad for AT.

**Comment by abbyad** [comment]
This feature will be included as part of 2.14 and will be (re)acceptance tested once there is a 2.14 beta release.

**Comment by sglangevin** [comment]
@abbyad per your comment, this is supposed to be released with 2.14.0. What is the status of AT?

**Comment by abbyad** [comment]
Having problems doing AT, likely because the API key and module ID are not being set in our build. @alxndrsn will take a look since those are needed to really make this feature available.

**Comment by garethbowen** [comment]
@alxndrsn Am I supposed to be code reviewing this? For a moment there I was assigned...

**Comment by alxndrsn** [comment]
Nope, slip of the mouse - no review required!

**Comment by abbyad** [comment]
Running [medic-android 0.4.0](https://github.com/medic/medic-android/releases/tag/v0.4.0) I still cannot launch the Simprints app. Identifying with the Simprints search button in people tab filter bar and the "Register with Simprints" button in a form gives the error: ``Uncaught Error: Java bridge method can't be invoked on a non-injected object`

## Identify
![image](https://user-images.githubusercontent.com/789512/33222282-fe21bf1a-d123-11e7-8aad-7670adc1f1ee.png)

## Register
![image](https://user-images.githubusercontent.com/789512/33222236-cbb59858-d123-11e7-99a8-2b9c72b80c3e.png)

**Comment by garethbowen** [comment]
Merged and backported to `2.14.x`

**Comment by abbyad** [comment]
Making progress! ...but still fails on `2.14.6-beta`.

The register button triggers Simprints app properly, but when handing back to the Medic app it appears as though no value is returned since the newly created contact has `simprints_id:""`. Unfortunately the only USB cable I have on hand today is power only so can't connect to see the console. I will report back with console output as soon as possible, latest by the end of this week.

**Comment by abbyad** [comment]
No output on console, but this is what we get on failed Registration on adb:
```
12-06 19:12:12.790  4718  4718 E Parcel  : Caused by: java.lang.ClassNotFoundException: Didn't find class "com.simprints.libsimprints.Registration" on path: DexPathList[[zip file "/data/app/org.medicmobile.webapp.mobile-1/base.apk"],nativeLibraryDirectories=[/data/app/org.medicmobile.webapp.mobile-1/lib/arm, /data/app/org.medicmobile.webapp.mobile-1/base.apk!/lib/armeabi-v7a, /system/lib, /vendor/lib]]
12-06 19:12:12.792  4718  4718 W MedicMobile:   at org.medicmobile.webapp.mobile.q.a(SimprintsSupport.java:90)
12-06 19:12:12.792  4718  4718 W MedicMobile:   at org.medicmobile.webapp.mobile.EmbeddedBrowserActivity.onActivityResult(EmbeddedBrowserActivity.java:159)
```

**Comment by abbyad** [comment]
Works well enough to demonstrate it working, although found a couple of small issues. We should eventually get to all of these, but perhaps not worth holding back the release for them unless they can be resolved quickly. Here they are, with the highest priority at the top:
- [ ]  If the Simprints search returns no one, all places and people are shown as the result. Would be more appropriate to say "No one found".
- [ ] When registering someone with `Register with Simprints` in a form there is no indication that the registration succeeded; the register button remains the same as before. For the time being we can we could have an Enketo _note_ saying "Simprints registration complete" beneath when the field has a value. We can then revisit this with input from @diannakane and designers when needed.
- [ ] Pressing back from the Simprints app goes to the Medic app, then goes back one page in the app. If Registering from a new person form, the Back in the Simprints app tries to exit the form. 

Any of these that are not dealt with in this issue should have separate issues created for them to be dealt with in due time and not lost in this issue.

**Comment by alxndrsn** [comment]
1. Investigating
2. Fixed
3. In conversation with Simprints about Simprints ID's handling of back presses

**Comment by abbyad** [comment]
Thanks @alxndrsn, the most recent fixes work well in `2.14.0-beta.7`. The back button behaviour seems related to the Simprints app as you guessed since it propagates the back button press from the _Bluetooth Error_, and the _Scanner Not Found_ pages, but not the _Why did you skip fingerprinting?_ page. It is currently being looked at by their team so this issue is moving over to Ready.
